src/Geometry/Primitives/AsphericSurface.jl

```diff
@@ -0,0 +1,341 @@
+    end
+
+end
+
+
+asphericType(z::AsphericSurface{T,3,Q,EVEN}) where {T<:Real,Q} = EVEN
+
+asphericType(z::AsphericSurface{T,3,Q,CONIC}) where {T<:Real,Q} = CONIC
+
+asphericType(z::AsphericSurface{T,3,Q,ODD}) where {T<:Real,Q} = ODD
+
+asphericType(z::AsphericSurface{T,3,Q,ODDEVEN}) where {T<:Real,Q} = ODDEVEN
+
+"""
+
+```julia
+EvenAsphericSurface(semidiameter, curvature::T, conic::T, aspherics::Vector{T}; normradius::T=semidiameter)
+```
+
+Surface incorporating an aspheric polynomial - radius, conic and aspherics are defined relative to absolute semi-diameter.
+
+`aspherics` should be an array of the even coefficients of the aspheric polynomial
+
+
+"""
+
+"""
+
+```julia
+OddAsphericSurface(semidiameter, curvature::T, conic::T, aspherics::Vector{T}; normradius::T=semidiameter)
+```
+
+Surface incorporating an aspheric polynomial - radius, conic and aspherics are defined relative to absolute semi-diameter.
+
+`aspherics`  should be an array of the odd coefficients of the aspheric polynomial starting with A1
+
+"""
+
+"""
+
+```julia
+OddEvenAsphericSurface(semidiameter, curvature::T, conic::T, aspherics::Vector{T}; normradius::T=semidiameter)
+```
+
+Surface incorporating an aspheric polynomial - radius, conic and aspherics are defined relative to absolute semi-diameter.
+
+`aspherics` should be an array of the both odd and even coefficients of the aspheric polynomial starting with A1
+
+"""
+
+
+export AsphericSurface, EvenAsphericSurface, OddAsphericSurface, OddEvenAsphericSurface
+
+
+uvrange(::AsphericSurface{T,N}) where {T<:Real,N} = ((zero(T), one(T)), (-T(π), T(π))) # ρ and ϕ
+
+semidiameter(z::AsphericSurface{T}) where {T<:Real} = z.semidiameter
+halfsizeu(z::AsphericSurface{T}) where {T<:Real} = semidiameter(z)
+halfsizev(z::AsphericSurface{T}) where {T<:Real} = semidiameter(z)
+
+boundingobj(z::AsphericSurface{T}) where {T<:Real} = z.boundingcylinder
+
+prod_step(z::AsphericSurface{T,N,Q,ODDEVEN}, r, r2) where {T<:Real,N,Q} = r, r
+prod_step(z::AsphericSurface{T,N,Q,ODD}, r, r2) where {T<:Real,N,Q} = r, r2
+prod_step(z::AsphericSurface{T,N,Q,EVEN}, r, r2) where {T<:Real,N,Q} = r2, r2
+
+function point(z::AsphericSurface{T,3,Q,M}, ρ::T, ϕ::T)::SVector{3,T} where {T<:Real,Q,M}
+    rad = z.semidiameter
+    r = ρ * rad
+    r2 = r^2
+    t = one(T) - z.curvature^2 * (z.conic + one(T)) * r^2
+    if t < zero(T)
+        return SVector{3,T}(NaN, NaN, NaN)
+    end
+    h = z.curvature * r2 / (one(T) + sqrt(t))
+   # sum aspheric
+    if M != CONIC
+        prod, step = prod_step(z, r, r2)  #multiple dispatch on R
+        asp,rest = Iterators.peel(z.aspherics)
+        h += asp * prod
+        for asp in rest
+            prod *= step
+            h += asp * prod
+        end
+    end
+    return SVector{3,T}(r * cos(ϕ), r * sin(ϕ), h)
+end
+
+partial_prod_step(z::AsphericSurface{T, 3, Q, EVEN}) where {T<:Real,Q} = r, r2, 2:2:2Q
```

Comment:
```suggestion
partial_prod_step(z::AsphericSurface{T,3,Q,EVEN}, r: T, r2: T) where {T<:Real,Q} = r, r2, 2:2:2Q
```
You're missing two arguments from the definition, also we tend not to use spaces in type refs